core/unit_test/TestGraph.hpp

```diff
@@ -420,112 +420,243 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), empty_graph) {
+
+#if defined(KOKKOS_ENABLE_CUDA)
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::Cuda>) {
+      using driver_t = Kokkos::Impl::ParallelFor<
+          functor_t,
+          Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::Impl::IsGraphKernelTag>,
+          Kokkos::Cuda>;
+      static_assert(
+          Kokkos::Impl::DeduceCudaLaunchMechanism<driver_t>::launch_mechanism ==
+          launch_mechanism);
+    }
+#endif
+
+#if defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::HIP>) {
+      using driver_t = Kokkos::Impl::ParallelFor<
+          functor_t,
+          Kokkos::RangePolicy<Kokkos::HIP, Kokkos::Impl::IsGraphKernelTag>,
+          Kokkos::HIP>;
+      static_assert(
+          Kokkos::Impl::DeduceHIPLaunchMechanism<driver_t>::launch_mechanism ==
+          launch_mechanism);
+    }
+#endif
+
+    const std::string kernel_name = "Let's make it a huge kernel";
+    const std::string alloc_label =
+        kernel_name + " - GraphNodeKernel global memory functor storage";
+
+    view_t data(Kokkos::view_alloc("witness", exec));
+
+    using namespace Kokkos::Test::Tools;
+    listen_tool_events(Config::DisableAll(), Config::EnableAllocs());
+
+    std::optional<Kokkos::Experimental::Graph<ExecSpace>> graph;
+
+    const void* ptr   = nullptr;
+    uint64_t ptr_size = 0;
+
+    const size_t range_end = 10;
+
+    auto lam_create = [&]() {
+      graph = Kokkos::Experimental::create_graph(exec, [&](const auto& root) {
+        auto node = root.then_parallel_for(
+            kernel_name, Kokkos::RangePolicy<ExecSpace>(0, range_end),
+            functor_t(data));
+      });
+    };
+
+    auto matcher_create = [&](AllocateDataEvent alloc) {
+      if (alloc.name != alloc_label)
         return MatchDiagnostic{
-            false, {"Either the name or pointer or size did not match"}};
-      }));
+            false, {"Allocation name mismatch (got " + alloc.name + ')'}};
+      if (alloc.size < sizeof(functor_t))
+        return MatchDiagnostic{
+            false,
+            {"Allocation size mismatch (expected at least " +
+             std::to_string(sizeof(functor_t)) + " but got " +
+             std::to_string(alloc.size) + ')'}};
+      ptr      = alloc.ptr;
+      ptr_size = alloc.size;
+      return MatchDiagnostic{true};
+    };
+
+    if constexpr (allocates_driver_memory_buffer) {
+      ASSERT_TRUE(
+          validate_existence(std::move(lam_create), std::move(matcher_create)));
+    } else {
+      ASSERT_TRUE(
+          validate_absence(std::move(lam_create), std::move(matcher_create)));
+    }
 
-  listen_tool_events(Config::DisableAll());
+    EXPECT_TRUE(static_cast<bool>(graph));
+    graph->instantiate();  // NOLINT(bugprone-unchecked-optional-access)
+
+    // Fencing the default execution space instance, as the node policy
+    // was created without giving an instance (it used the default one).
+    ExecSpace{}.fence(
```

Comment:
The function takes an exec argument
```suggestion
    exec.fence(
```

Reply:
I think that it is right to use the default exec here. Github is coloring this line of code and the comment right above, but these lines have only been moved a bit and they were already in the test before this PR.

The way the test was and still is set up is that the test passes a policy with a default exec to `then_parallel_for`. And so, when `Kokkos` chooses the global launch mechanism, the impact of passing such a policy is that `Kokkos` will use the default exec to deep copy the driver to the device memory. And so as the comment says, we must fence the default exec to make sure that this deep copy is finished.